Allow returning custom metrics from ConnectorPageSource

rzeyde-varada opened the pull request:
We suggest the following simple mechanism for returning and aggregating custom per-query metrics from each page source:
https://github.com/prestosql/presto/compare/master...rzeyde-varada:custom-metrics?expand=1#diff-1b3e5d066c2bad4b806846b761b9d68526c7abb942df89f1b6b92a8e76fe2948

For simplicity, each ConnectorPageSource returns a single mapping from strings to longs, and the engine aggregates them across all splits. This allows the connectors to expose various "low-level" counters without changing modifying the engine code (to return them via `OperatorStats`).

I would be happy to hear your opinion about this feature - any comments/questions/suggestions are welcome :)

sopel39 commented:
@rzeyde-varada Could you describe how you use such metrics?

rzeyde-varada commented:
We use them for per-query performance analysis (e.g. I/O-related statistics) by exposing low-level counters from our storage engine.

sopel39 commented on presto-main/src/main/java/io/prestosql/operator/CustomMetrics.java:
```diff
@@ -0,0 +1,46 @@
+import java.util.HashMap;
+import java.util.Map;
+
+public class CustomMetrics
```
this object should be immutable

rzeyde-varada replied:
Fixed.

sopel39 commented on presto-main/src/main/java/io/prestosql/operator/CustomMetrics.java:
```diff
@@ -0,0 +1,46 @@
+        metrics = new HashMap<>();
+    }
+
+    public synchronized void set(Map<String, Long> metrics)
```
this should return new object

rzeyde-varada replied:
Refactored.

sopel39 commented on presto-main/src/main/java/io/prestosql/operator/ScanFilterAndProjectOperator.java:
```diff
@@ -76,6 +77,7 @@
     private long physicalBytes;
     private long readTimeNanos;
     private long dynamicFilterSplitsProcessed;
+    private Map<String, Long> customMetrics = Map.of();
```
`ImmutableMap.of()`

sopel39 commented on presto-main/src/main/java/io/prestosql/operator/OperatorContext.java:
```diff
@@ -218,6 +220,14 @@ public void recordDynamicFilterSplitProcessed(long dynamicFilterSplits)
         dynamicFilterSplitsProcessed.getAndAdd(dynamicFilterSplits);
     }
 
+    /** Overwrites the metrics with the latest one.
```
use correct javadoc formatting

rzeyde-varada replied:
Fixed.

sopel39 reviewed (commented):
Initial pass. I think the idea is fine. What do you think @dain @martint ?

electrum commented:
Some design questions:

* Are the metric keys expected to be dynamic? Would it be too limiting for a connector to statically register the list of metrics it wants to expose?
* What about types that are not a `long`, such as percentile distribution, HyperLogLog, etc.? We'd need to know how to serialize and merge them.

electrum commented:
Looking at `OperatorStats` today, it appears that all of the metrics could be stored as a `long`, but would require more than just addition for merging, including floating point addition and maximum.

rzeyde-varada commented:
> Would it be too limiting for a connector to statically register the list of metrics it wants to expose?

Statically registering the list of metrics sounds good.
Is there an existing example for such API that we can (re-)use here?

> * What about types that are not a `long`, such as percentile distribution, HyperLogLog, etc.? We'd need to know how to serialize and merge them.

We have used `long` values (merged by addition) since it's the simplest method and works for most of our statistics.
I agree that adding floating point and maximum (for merging) can help - but wasn't sure about the extra complexity...

rzeyde-varada commented:
> Initial pass. I think the idea is fine. 

Thanks!
Updated the code to address the comments above.

rzeyde-varada commented:
Updated the code to support multiple metrics' types (e.g. `long`/`double`) and aggregation (e.g. `sum`/`max`).
WDYT?

rzeyde-varada commented:
Ping :)

rzeyde-varada commented:
I have updated the PR, and I think that the existing design should work for simple counters, gauges and histograms which use a single type of aggregation (e.g. sum/max).

I’ve also added a simple `Histogram` class for exposing and aggregating histograms via custom metrics (similar to https://prometheus.io/docs/concepts/metric_types/#histogram).

rzeyde-varada commented:
Rebased (to resolve a small merge conflict).

martint commented on core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorPageSource.java:
```diff
@@ -70,4 +72,9 @@ void close()
         return NOT_BLOCKED;
     }
+
+    default Map<String, Metric> getCustomMetrics()
```
Call this `getMetrics()`.

Also, it needs to document the semantics of the returned metrics -- i.e., they are a new snapshot of the metrics, metrics need to be immutable, etc.

martint commented on core/trino-spi/src/main/java/io/trino/spi/metrics/CustomMetrics.java:
```diff
@@ -0,0 +1,45 @@
+        this(Arrays.asList(maps));
+    }
+
+    public CustomMetrics(Iterable<Map<String, Metric>> maps)
```
What's the use case for taking multiple maps?

rzeyde-varada replied:
I am using it to merge multiple custom metrics at `OperatorStats#add`, but it's probably not a good choice for a c-tor - so I have renamed it into a static method `Metrics#merge`: https://github.com/trinodb/trino/pull/6232/commits/b4f3406d2b18f7fe821baeae0d1cec3156b89700#diff-de6c1298bb6eabc28a83bbb6305b6de4f2900681b3dfab627a0d541e2e50230aR34

martint commented on core/trino-spi/src/main/java/io/trino/spi/metrics/LongSumValue.java:
```diff
@@ -0,0 +1,50 @@
+    }
+
+    @Override
+    public Metric merge(Metric other)
```
Take a look at `io.trino.util.Mergeable` and implementations of that. We may want to do something similar. Or move Mergeable to the SPI for use in these.

rzeyde-varada replied:
Sounds good - I have moved `Mergeable` to `trino-spi` at https://github.com/trinodb/trino/pull/6232/commits/ad3f94ae9739213837a61877d38e430151af18fd and refactored the merging code following the code at `OperatorStats#mergeInfo`.

rzeyde-varada commented on core/trino-spi/src/main/java/io/trino/spi/metrics/Histogram.java:
```diff
@@ -0,0 +1,154 @@
+
+import java.util.Arrays;
+
+public class Histogram
```
IIUC, adding airlift's `stats` to `trino-spi` seems to require adding even more dependencies to the SPI - e.g. `guava` (which is currently is used only for tests). However, after adding `guava` as a `provided` dependency [1] I got the following errors when building `trino-tpch`:
```
$ ./mvnw install -Dmaven.gitcommitid.skip=true -DskipTests -pl plugin/trino-tpch -am 
...
[ERROR] Failed to execute goal io.trino:trino-maven-plugin:10:check-spi-dependencies (default-check-spi-dependencies) on project trino-tpch: 
[ERROR] 
[ERROR] Trino plugin dependency org.hdrhistogram:HdrHistogram must not have scope 'provided'. It is not part of the SPI and will not be available at runtime.
```

Since `org.hdrhistogram` is an airlift `stats` dependency [2], I tried to add the following fixes to `trino-tpch` [3], but it seems that I didn't specified the dependency correctly :(

I would appreciate if you could please help me to find the correct way of adding the `TDigest` to `trino-spi`...

[1] https://github.com/rzeyde-varada/prestosql/commit/5b80288ec8bfc628bfc7071e0784877f6fa0c094
[2] https://github.com/airlift/airlift/blob/1e50fbd21c43ea355b7415af6a39ca7ad42c1daa/stats/pom.xml#L20
[3] https://github.com/rzeyde-varada/prestosql/commit/d1228f30554442a240517344e9cedd1051befbbe

rzeyde-varada replied:
Thanks!
I think that I will move the implementation into a toolkit.

rzeyde-varada replied:
Done in https://github.com/trinodb/trino/pull/6232/commits/b50d4735e4da4d736b632f8dd0e7c9bf0e50db2e.

rzeyde-varada commented:
Thanks for the review!
I have resolved most of the issues - but am not sure about https://github.com/trinodb/trino/pull/6232#discussion_r625352711, due to dependency errors I am getting after adding airlift `stats` to the SPI... could you please take a look?

rzeyde-varada commented:
I have split the `Metric` interface (cff73a4) from its implementations (b50d473).

raunaqmorarka commented on core/trino-main/src/main/java/io/trino/operator/OperatorStats.java:
```diff
@@ -169,6 +176,7 @@ public OperatorStats(
         this.outputPositions = outputPositions;
 
         this.dynamicFilterSplitsProcessed = dynamicFilterSplitsProcessed;
+        this.customMetrics = customMetrics;
```
requireNonNull

rzeyde-varada replied:
Thanks, fixed.

raunaqmorarka commented on core/trino-spi/src/main/java/io/trino/spi/metrics/Metrics.java:
```diff
@@ -0,0 +1,57 @@
+        this.map = Map.copyOf(requireNonNull(map, "map is null"));
+    }
+
+    public static Metrics merge(Iterable<Map<String, Metric>> maps)
```
Can we make it
`Map<String, Metric> merge(Iterable<Map<String, Metric>> metrics)`
or
`Metrics merge(Iterable<Metrics> metrics)` ?
Looks inconsistent to use the wrapper class sometimes and the Map directly in other places

rzeyde-varada replied:
Good point - changed to use `Metrics` instead of `Map`.

raunaqmorarka commented on core/trino-main/src/main/java/io/trino/operator/OperatorStats.java:
```diff
@@ -332,6 +340,16 @@ public long getDynamicFilterSplitsProcessed()
+    @JsonProperty
+    public Map<String, Metric> getCustomMetrics()
+    {
+        if (customMetrics == null) {
```
null handling shouldn't be needed

raunaqmorarka commented on core/trino-main/src/main/java/io/trino/operator/OperatorStats.java:
```diff
@@ -115,6 +121,7 @@ public OperatorStats(
             @JsonProperty("outputPositions") long outputPositions,
 
             @JsonProperty("dynamicFilterSplitsProcessed") long dynamicFilterSplitsProcessed,
+            @JsonProperty("customMetrics") Map<String, Metric> customMetrics,
```
Could we use `Metrics` instead of `Map<String, Metric>` ?
Or if we just want to use maybe `Map<String, Metric>` directly, we can consider getting rid of Metrics wrapper and put the merge method in some utils class.

rzeyde-varada replied:
Sounds good, changed to use `Metrics` instead of `Map`.

raunaqmorarka commented on core/trino-main/src/main/java/io/trino/operator/WorkProcessorPipelineSourceOperator.java:
```diff
@@ -675,6 +681,7 @@ public void close()
         final AtomicLong outputPositions = new AtomicLong();
 
         final AtomicLong dynamicFilterSplitsProcessed = new AtomicLong();
+        final AtomicReference<Metrics> customMetrics = new AtomicReference(Metrics.EMPTY);
```
`new AtomicReference<>`

rzeyde-varada replied:
Thanks, fixed.

raunaqmorarka commented on core/trino-main/src/main/java/io/trino/operator/OperatorContext.java:
```diff
@@ -82,6 +83,7 @@
     private final CounterStat outputPositions = new CounterStat();
 
     private final AtomicLong dynamicFilterSplitsProcessed = new AtomicLong();
+    private final AtomicReference<Metrics> customMetrics = new AtomicReference(Metrics.EMPTY);  // this is not incremental, but gets overwritten by the latest value.
```
`new AtomicReference<>`

rzeyde-varada replied:
Thanks, fixed.

raunaqmorarka commented on core/trino-main/src/main/java/io/trino/operator/OperatorStats.java:
```diff
@@ -451,6 +469,8 @@ public OperatorStats add(Iterable<OperatorStats> operators)
         long outputPositions = this.outputPositions;
 
         long dynamicFilterSplitsProcessed = this.dynamicFilterSplitsProcessed;
+        ImmutableList.Builder<Map<String, Metric>> metricsMaps = ImmutableList.builder();
```
Do we need a list ?
Could we merge the maps one at a time like the other metrics here ?

rzeyde-varada replied:
Sounds good, changed to merge the metrics on at a time.

rzeyde-varada commented on core/trino-main/src/main/java/io/trino/operator/WorkProcessorSourceOperatorAdapter.java:
```diff
@@ -228,6 +232,11 @@ private void updateOperatorStats()
             previousDynamicFilterSplitsProcessed = currentDynamicFilterSplitsProcessed;
         }
+
+        if (!currentMetrics.equals(previousMetrics)) {
```
Thanks, fixed to use `!=`.

rzeyde-varada commented on core/trino-spi/src/main/java/io/trino/spi/metrics/Metrics.java:
```diff
@@ -0,0 +1,57 @@
+
+    public static Metrics merge(Iterable<Map<String, Metric>> maps)
+    {
+        Map<String, Metric> merged = new HashMap();
```
Thanks, fixed.

rzeyde-varada commented on core/trino-spi/src/main/java/io/trino/spi/metrics/Metrics.java:
```diff
@@ -0,0 +1,57 @@
+                .forEach(entry -> merged.merge(
+                        entry.getKey(),
+                        entry.getValue(),
+                        (left, right) -> mergeInternal(left, right))));
```
Thanks, fixed.

rzeyde-varada commented on core/trino-spi/src/main/java/io/trino/spi/metrics/Metrics.java:
```diff
@@ -0,0 +1,57 @@
+{
+    public static final Metrics EMPTY = new Metrics(Map.of());
+
+    private final Map<String, Metric> map;
```
Good point - changed `Metric` to extend `Mergeable`.

rzeyde-varada commented on core/trino-spi/src/main/java/io/trino/spi/metrics/Metrics.java:
```diff
@@ -0,0 +1,57 @@
+{
+    public static final Metrics EMPTY = new Metrics(Map.of());
+
+    private final Map<String, Metric> map;
```
Thanks, fixed.

rzeyde-varada commented:
Many thanks for the review!
I have updated the PR, please take a look :)

raunaqmorarka commented on core/trino-spi/src/main/java/io/trino/spi/metrics/Metrics.java:
```diff
@@ -0,0 +1,56 @@
+        this.metrics = Map.copyOf(requireNonNull(metrics, "metrics is null"));
+    }
+
+    public static Metrics merge(Metrics... metrics)
```
Since we only give two metrics as input, it might be better to simplify this method to just take 2 metrics as input.
Or you could even make this class implement `Mergeable<Metrics>` interface.

rzeyde-varada replied:
Made `Metrics` mergeable.

raunaqmorarka commented on core/trino-main/src/main/java/io/trino/operator/OperatorStats.java:
```diff
@@ -498,6 +509,7 @@ public OperatorStats add(Iterable<OperatorStats> operators)
             outputPositions += operator.getOutputPositions();
 
             dynamicFilterSplitsProcessed += operator.getDynamicFilterSplitsProcessed();
+            metrics = Metrics.merge(metrics, operator.getMetrics());
```
I think one disadvantage of working with Metrics wrapper is that it would result in the underlying map getting copied many times here.
It might be more efficient to work directly with the underlying Map for merging here
E.g.
```
Map<> metricsAccumulator = this.getMetrics().getMetrics()
...
for (OperatorStats operator : operators) {
....
    operator.getMetrics().forEach((key, value) -> metricsAccumulator.merge(key, value, (m1, m2) -> m1.mergeWith(m2))));
....
}
Metrics merged = new Metrics(metricsAccumulator);

```

rzeyde-varada replied:
Sounds good - created an helper class for accumulating metrics (similar to the "builder" pattern).

raunaqmorarka commented on core/trino-main/src/main/java/io/trino/operator/WorkProcessorSourceOperatorAdapter.java:
```diff
@@ -175,6 +177,7 @@ public void close()
             throws Exception
     {
         sourceOperator.close();
+        operatorContext.setLatestMetrics(sourceOperator.getConnectorMetrics());
```
Would it better to collect `sourceOperator.getConnectorMetrics()` before `sourceOperator.close()` as sourceOperator could decide to clear it's metrics to release resources on `close` ?

rzeyde-varada replied:
Sounds good - moved collections before closing the operator.

raunaqmorarka reviewed (commented):
Minor comments, LGTM overall

rzeyde-varada commented on plugin/trino-memory/src/test/java/io/trino/plugin/memory/TestMemorySmoke.java:
```diff
@@ -99,6 +104,52 @@ public void testSelect()
+                ((Count) metrics.get("closed")).getValue());
+        assertGreaterThan(
+                ((Count) metrics.get("closed")).getValue(),
+                0L);
```
Fixed the first two, but the last one still requires comparing the underlying count (since `Metric` is not `Comparable`) IIUC.

rzeyde-varada commented on plugin/trino-memory/src/test/java/io/trino/plugin/memory/TestMemorySmoke.java:
```diff
@@ -99,6 +104,52 @@ public void testSelect()
+                .getQueryStats()
+                .getOperatorSummaries()
+                .stream()
+                .map(s -> s.getMetrics())
```
Good catch, thanks!

rzeyde-varada commented on lib/trino-plugin-toolkit/src/test/java/io/trino/plugin/base/metrics/TestMetrics.java:
```diff
@@ -0,0 +1,67 @@
+import io.trino.spi.metrics.Metrics;
+import org.testng.annotations.Test;
+
+import static org.testng.Assert.assertEquals;
```
Fixed.

rzeyde-varada commented:
Many thanks for the review!
The PR is updated.

martint commented on core/trino-spi/src/main/java/io/trino/spi/metrics/Metrics.java:
```diff
@@ -0,0 +1,80 @@
+            return new Metrics(merged);
+        }
+
+        private static Metric mergeInternal(Metric left, Metric right)
```
This method seems unnecessary. It's used only once and it's fairly trivial. Just inline it where it's used.

rzeyde-varada replied:
Sounds good - done.

martint commented on core/trino-spi/src/main/java/io/trino/spi/metrics/Metrics.java:
```diff
@@ -0,0 +1,80 @@
+    @JsonCreator
+    public Metrics(Map<String, Metric> metrics)
+    {
+        this.metrics = requireNonNull(metrics, "metrics is null");
```
This class is holding on to the map provided by the caller. That can cause potential problems down the road if the map is modified (especially in a non-thread-safe manner). Do `Map.copyOf(...)`

rzeyde-varada replied:
Thanks, fixed.

martint commented on core/trino-spi/src/main/java/io/trino/spi/metrics/Metrics.java:
```diff
@@ -0,0 +1,80 @@
+        this.metrics = requireNonNull(metrics, "metrics is null");
+    }
+
+    @JsonProperty("metrics")
```
Use `@JsonValue` so that it doesn't introduce an artificial nested JSON object.

rzeyde-varada replied:
Thanks, fixed.

martint commented on lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/metrics/Histogram.java:
```diff
@@ -0,0 +1,46 @@
+    }
+
+    @JsonProperty("digest")
+    public TDigest getDigest()
```
TDigest is not json-serializable.

rzeyde-varada replied:
Oops, you're right - added a custom converter to serialize/deserialize TDigest (using base64).

martint reviewed (commented):
One issue with this implementation is that there's nothing the engine can do with the metrics other than serialize them to JSON. The implementation of a metric is completely opaque to the engine.

I think what we want is to define interfaces for a set of well-known metric types (Count, Distribution) with known methods such as `getTotal()`, `getPercentiles()`, `merge()`, etc., similar to what we had before. But, leave the implementation (whether it uses t-digest, qdigest, etc) to the provider, since we can't have those data structures in the SPI.

rzeyde-varada commented on core/trino-main/src/main/java/io/trino/operator/WorkProcessorPipelineSourceOperator.java:
```diff
@@ -675,6 +681,7 @@ public void close()
         final AtomicLong outputPositions = new AtomicLong();
 
         final AtomicLong dynamicFilterSplitsProcessed = new AtomicLong();
+        final AtomicReference<Metrics> customMetrics = new AtomicReference<>(Metrics.EMPTY);
```
Renamed to `connectorMetrics`.

rzeyde-varada commented:
Many thanks for the reivew - I've updated the PR, but I have a few questions about the suggestion above:

> One issue with this implementation is that there's nothing the engine can do with the metrics other than serialize them to JSON. The implementation of a metric is completely opaque to the engine.

I agree - the original motivation for this was as following:
- our metrics' analysis is happening outside of the Trino query execution engine, so we currently use the engine only to merge and serialize metrics.
- following https://github.com/trinodb/trino/pull/6232#issuecomment-742144191, we didn't want to expose the data type of the metric (e.g. `long`, `double`) from the SPI.

> I think what we want is to define interfaces for a set of well-known metric types (Count, Distribution) with known methods such as `getTotal()`, `getPercentiles()`, `merge()`, etc., similar to what we had before. But, leave the implementation (whether it uses t-digest, qdigest, etc) to the provider, since we can't have those data structures in the SPI.

There are two small issues I am not sure about:
1. Should we have separate interface and implementation for `Count` metric? It looks quite simple, so maybe we can just implement it in the SPI?
2. Just to be sure: should `Count` and `Distribution` be separate interfaces, both extending the `Metric` interface?

rzeyde-varada commented:
> * our metrics' analysis is happening outside of the Trino query execution engine, so we currently use the engine only to merge and serialize metrics.

BTW, could you please elaborate how the new metric interfaces would be used by the engine?

rzeyde-varada commented:
Updated the PR with https://github.com/trinodb/trino/pull/6232/commits/e1bcb1c03f3c895bf2f8a062143e3c7006ad3fb1 - please let me know if it is the correct way to introduce the interfaces.

martint commented on core/trino-main/src/main/java/io/trino/operator/WorkProcessorSourceOperatorAdapter.java:
```diff
@@ -228,6 +232,11 @@ private void updateOperatorStats()
             previousDynamicFilterSplitsProcessed = currentDynamicFilterSplitsProcessed;
         }
+
+        if (currentMetrics != previousMetrics) {
+            operatorContext.setLatestMetrics(currentMetrics);
+            previousMetrics = currentMetrics;
```
Why not set it unconditionally?

rzeyde-varada replied:
It can be done, but I thought to optimize for the case where `currentMetrics == Metrics.EMPTY` (which currently happens for all connectors that don't implement `ConnectorPageSource#getMetrics`).

martint replied:
setLatestMetrics just sets a field. There's no processing involved, so I don't think this optimization really optimizes much. We can add it in the future if we observe that being an actual problem.

martint commented on core/trino-spi/src/main/java/io/trino/spi/metrics/Count.java:
```diff
@@ -0,0 +1,19 @@
+
+public interface Count
+{
+    double getTotal();
```
Why `double`?

rzeyde-varada replied:
In case a connector would like to aggregate a floating-point metric, returning `double` should work both for integer and floating-point metrics.
Should I change it to `long`?

martint replied:
I think that would be a different type of metric, though. Let's change it to `long`. We can reconsider in the future if we find an actual use case for it.

martint commented on core/trino-spi/src/main/java/io/trino/spi/metrics/Count.java:
```diff
@@ -0,0 +1,19 @@
+ */
+package io.trino.spi.metrics;
+
+public interface Count
```
Did you try making Count and Histogram extend from Metric?

rzeyde-varada replied:
Good point, thanks!
Fixed in 749ef4f.

martint commented on core/trino-spi/src/main/java/io/trino/spi/metrics/Distribution.java:
```diff
@@ -0,0 +1,21 @@
+
+public interface Distribution
+{
+    double getTotal();
```
Why `double`?

rzeyde-varada replied:
Because `TDigest#getCount` returns `double` - so I used the same type here.

martint replied:
That's because TDigest and QDigest are designed to support exponential decay, which results in fractional weights over time. I don't think we need that here.

rzeyde-varada commented:
Thanks for the review!
I've updated the PR and responded to the comments above.

rzeyde-varada commented:
[`TestOraclePoolConnectorSmokeTest` failure](https://github.com/trinodb/trino/pull/6232/checks?check_run_id=2870094940) seems to be unrelated to this PR:
```
java.lang.RuntimeException: java.sql.SQLRecoverableException: IO Error: Got minus one from a read call
	at io.trino.plugin.oracle.TestingOracleServer.<init>(TestingOracleServer.java:81)
	at io.trino.plugin.oracle.TestOraclePoolConnectorSmokeTest.createQueryRunner(TestOraclePoolConnectorSmokeTest.java:35)
	at io.trino.testing.AbstractTestQueryFramework.init(AbstractTestQueryFramework.java:91)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:104)
	at org.testng.internal.Invoker.invokeConfigurationMethod(Invoker.java:515)
	at org.testng.internal.Invoker.invokeConfigurations(Invoker.java:217)
	at org.testng.internal.Invoker.invokeConfigurations(Invoker.java:144)
	at org.testng.internal.TestMethodWorker.invokeBeforeClassMethods(TestMethodWorker.java:169)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:108)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.sql.SQLRecoverableException: IO Error: Got minus one from a read call
	at oracle.jdbc.driver.T4CConnection.logon(T4CConnection.java:854)
	at oracle.jdbc.driver.PhysicalConnection.connect(PhysicalConnection.java:793)
	at oracle.jdbc.driver.T4CDriverExtension.getConnection(T4CDriverExtension.java:57)
	at oracle.jdbc.driver.OracleDriver.connect(OracleDriver.java:747)
	at oracle.jdbc.driver.OracleDriver.connect(OracleDriver.java:562)
	at io.trino.plugin.jdbc.DriverConnectionFactory.openConnection(DriverConnectionFactory.java:65)
	at io.trino.plugin.jdbc.RetryingConnectionFactory.lambda$openConnection$0(RetryingConnectionFactory.java:54)
	at net.jodah.failsafe.Functions.lambda$get$0(Functions.java:48)
	at net.jodah.failsafe.RetryPolicyExecutor.lambda$supply$0(RetryPolicyExecutor.java:62)
	at net.jodah.failsafe.Execution.executeSync(Execution.java:129)
	at net.jodah.failsafe.FailsafeExecutor.call(FailsafeExecutor.java:376)
	at net.jodah.failsafe.FailsafeExecutor.get(FailsafeExecutor.java:67)
	at io.trino.plugin.jdbc.RetryingConnectionFactory.openConnection(RetryingConnectionFactory.java:54)
	at io.trino.plugin.oracle.TestingOracleServer.<init>(TestingOracleServer.java:73)
	... 15 more
Caused by: oracle.net.ns.NetException: Got minus one from a read call
	at oracle.net.ns.NSProtocolNIO.doSocketRead(NSProtocolNIO.java:557)
	at oracle.net.ns.NIOPacket.readNIOPacket(NIOPacket.java:403)
	at oracle.net.ns.NSProtocolNIO.negotiateConnection(NSProtocolNIO.java:127)
	at oracle.net.ns.NSProtocol.connect(NSProtocol.java:340)
	at oracle.jdbc.driver.T4CConnection.connect(T4CConnection.java:1596)
	at oracle.jdbc.driver.T4CConnection.logon(T4CConnection.java:588)
	... 28 more
```

rzeyde-varada commented:
Ping :)

rzeyde-varada commented:
Thanks for the review!
Fixed the comments and updated the PR :)

rzeyde-varada commented:
Many thanks!

findepi commented on plugin/trino-memory/src/test/java/io/trino/plugin/memory/TestMemorySmoke.java:
```diff
@@ -107,6 +113,40 @@ public void testSelect()
+    @Test
+    public void testCustomMetricsScanFilter()
+    {
+        Map<String, Metric> metrics = collectCustomMetrics("SELECT partkey FROM part WHERE partkey % 1000 > 0");
```
Would it be possible to avoid using raw types here?

rzeyde-varada replied:
Sounds good - do you mean returning a `Metrics` object from `collectCustomMetrics()` (instead of a `Map`)?
If so, I think it can be fixed by https://github.com/rzeyde-varada/trino/commit/7332dbeb4935f5c7781cfc23f4de4cba4a2975ca.

findepi replied:
I don't know this code, so cannot suggest something.
Just my IDE warns me that `Metric` is a generic class, but here it's used without `<...>`.

maybe `Metric<?>` would do for now

rzeyde-varada replied:
Opened #8707.
